lte/gateway/c/session_manager/SetMessageManagerHandler.cpp

```diff
@@ -72,50 +72,95 @@ void SetMessageManagerHandler::SetAmfSessionContext(
   // Print the message from AMF
   PrintGrpcMessage(static_cast<const google::protobuf::Message&>(request_cpy));
   response_callback(Status::OK, SmContextVoid());
-  /*The Event Based main_thread invocation and runs to handle session state*/
-  m5g_enforcer_->get_event_base().runInEventBaseThread(
-      [this, response_callback, request_cpy]() {
-        // extract values from proto
-        auto imsi = request_cpy.common_context().sid().id();
-        std::string dnn =
-            request_cpy.common_context().apn();  // may not required for demo-1
-        std::string session_id = id_gen_.gen_session_id(imsi);
-
-        MLOG(MDEBUG) << "Requested session from UE with IMSI: " << imsi
-                     << " Generated sessioncontext ID" << session_id;
-        /*reach complete message from proto message*/
-        SessionConfig cfg = m5g_build_session_config(request_cpy);
-
-        /*check if it's initial message*/
-        if ((cfg.rat_specific_context.m5gsm_session_context().rquest_type() ==
-             INITIAL_REQUEST) &&
-            (cfg.common_context.sm_session_state() == CREATING_0)) {
-          /* it's new UE establisment request and need to create the session
-           * context
-           */
-          MLOG(MDEBUG)
-              << "AMF request type INITIAL_REQUEST and session state CREATING";
-          /* Read the SessionMap from global session_store,
-           * if it is not found, it will be added w.r.t imsi
-           */
+
+  /* Read the proto message and check for state. Get the config out of proto.
+   * if state is CREATING - New session to be created
+   * if state is RELEASE  - The session to be delered
+   */
+  // Fetch complete message from proto message
+  SessionConfig cfg = m5g_build_session_config(request_cpy);
+  // extract values from proto
+  auto imsi       = request_cpy.common_context().sid().id();
+  std::string dnn = request_cpy.common_context().apn();
+  if (cfg.common_context.sm_session_state() == RELEASED_4) {
+    if (cfg.common_context.sm_session_version() != 0) {
+      /* This is a serious sync error, as AMF passing wrong version values
+       * for this session to release. As session version can never be 0.
+       * Print the error message and keep releasing the respective session
+       */
+      MLOG(MERROR) << "Wrong version received from AMF for IMSI " << imsi
+                   << " but continuing release request";
+    }
+    MLOG(MINFO) << "Release request for session from IMSI: " << imsi << " DNN "
+                << dnn;
+    // Requested message from AMF to release the session
+    m5g_enforcer_->get_event_base().runInEventBaseThread(
+        [this, cfg, imsi, dnn]() {
           auto session_map = session_store_.read_sessions({imsi});
-          send_create_session(session_map, imsi, session_id, cfg);
-        }
-      });
+          initiate_release_session(session_map, dnn, imsi);

```

Comment:
not sure how this works on 5g, but in LTE we can have several sessions. If 5g works the same, aren't we releasing all the sessions here?

Reply:
I'm pretty sure dnn=apn, so it should be unique per-APN. But generally we should be using session_id to identify everything unless we do not have the information.

Reply:
@oriol,  Currently we implemented -->IMSI session map entries deletion per session,  AMF sends "deleteion requet per session",  later we might implement IMSI level deregister functionality, as part of that all sessions belongs to that imsi are deleted.    @marie -->  session_id scope is limited to SMF, AMF doesn't aware of that.  Then only link we have "imsi+dnn" forms a unique session id.